### PR TITLE
Add ObjFactory for creation and management of collections

### PR DESCRIPTION
**Pull request by pbougue**
used in pt_data

### CODE REVIEW

File: obj_factory.h [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        ObjType* obj = vec.back().get();
+        obj->idx = vec.size() - 1;
+        obj->uri = uri;
+        map[obj->uri] = obj;
+        return obj;
+    }
+
+    void check_free_uri(const std::string& uri) {
+        if (navitia::contains(map, uri)) {
+            throw(std::logic_error(std::string("In ") + typeid(*this).name() +
```
Comment (antoine-de) [inline obj_factory.h]: why do you do this ?

File: obj_factory.h [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+template<typename ObjType>
+class ObjFactory {
+public:
+    typedef typename std::vector<std::unique_ptr<ObjType>>::iterator iterator;
+    typedef typename std::vector<std::unique_ptr<ObjType>>::const_iterator const_iterator;
+
+    template<typename ...Args>
+    ObjType* emplace(const std::string& uri, Args&& ...args) {
+        check_free_uri(uri);
+        vec.push_back(std::make_unique<ObjType>(std::forward<Args>(args)...));
```
Comment (antoine-de) [inline obj_factory.h]: why do you use push_back and not emplace_back ?

the mechanism does not seem  straight forward.
wouldn't it be better for you to create the unique_ptr and pass it to a `save` method (your `manage_vec_back_consistency`) ?

File: obj_factory.h [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    const_iterator begin() const {return std::begin(vec);}
+    iterator end() {return std::end(vec);}
+    const_iterator end() const {return std::end(vec);}
+
+    size_t size() const noexcept {return vec.size();}
+
+    template<class Archive> void serialize(Archive & ar, const unsigned int) { ar & vec & map; }
+
+private:
+    std::vector<std::unique_ptr<ObjType>> vec;
```
Comment (antoine-de) [inline obj_factory.h]: wouldn't it be better to use a IdxMap here ?

**Comment by antoine-de** [comment]
A PR without test is so sad... :cry:

File: obj_factory.h [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+https://groups.google.com/d/forum/navitia
+www.navitia.io
+*/
+#pragma once
+#include <vector>
+#include <unordered_map>
+#include "utils/functions.h"
+#include "utils/idx_map.h"
+
+namespace navitia{
```
Comment (antoine-de) [inline obj_factory.h]: spacing

File: obj_factory.h [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+/*
+ * Factory handling all collections for one object
+ * Object must inherit from navitia::type::Header
+ */
+template<typename ObjType>
+class ObjFactory {
+public:
+    typedef typename std::vector<std::unique_ptr<ObjType>>::iterator iterator;
```
Comment (antoine-de) [inline obj_factory.h]: since you use typedef, I would have typedefed the inner vector too

File: obj_factory.h [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    ObjType* get_mut(const Idx<ObjType>& idx) {
+        if (idx.val >= vec.size()) {
+            return nullptr;
+        }
+        return vec[idx.val];
+    }
+
+    iterator begin() {return std::begin(vec);}
```
Comment (antoine-de) [inline obj_factory.h]: spacing

File: obj_factory.h (hunk 5 above)
Comment (antoine-de) [inline obj_factory.h]: you should use `using` instead of `typedef`

File: obj_factory.h (hunk 3 above)
Comment (pbougue) [inline obj_factory.h]: Not the expected use of IdxMap, but it's a possibility.
It's not "mandatory" here I'd say.

File: obj_factory.h (hunk 1 above)
Comment (pbougue) [inline obj_factory.h]: To help debug and have an idea of the collection having the problem.

File: obj_factory.h (hunk 2 above)
Comment (pbougue) [inline obj_factory.h]: Totally agree for `emplace_back`, less convinced about this `save` method even it actually would be more straightforward.

File: obj_factory.h [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    size_t size() const noexcept {return vec.size();}
+
+    template<class Archive> void serialize(Archive & ar, const unsigned int) { ar & vec & map; }
+
+private:
+    std::vector<std::unique_ptr<ObjType>> vec;
+    std::unordered_map<std::string, ObjType*> map;
+
+    ObjType* manage_vec_back_consistency(const std::string& uri) {
+        ObjType* obj = vec.back().get();
```
Comment (xlqian) [inline obj_factory.h]: `ObjType& obj = vec.back();` ?

File: obj_factory.h [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        ObjType* obj = get_mut(uri);
+        if (obj == nullptr) {
+            obj = emplace(uri, args...);
+        }
+        return obj;
+    }
+
+    ObjType* insert(const std::string& uri, ObjType&& obj) {
+        check_free_uri(uri);
+        vec.push_back(std::make_unique<ObjType>(std::move(obj)));
```
Comment (xlqian) [inline obj_factory.h]: Here so `insert()` takes a rvalue ONLY?

File: obj_factory.h (hunk 1 above)
Comment (xlqian) [inline obj_factory.h]: Log the error

File: obj_factory.h (hunk 8 above)
Comment (pbougue) [inline obj_factory.h]: yes, you give the property of the obj to obj_factory

**Comment by pbougue** [comment]
yep I hope you'll do something about it as soon as possible!

File: obj_factory.h [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+        obj->uri = uri;
+        map[obj->uri] = obj;
+        return obj;
+    }
+
+    template<typename ...Args>
+    ObjType* get_or_create(const std::string& uri, Args&& ...args) {
+        ObjType* obj = get_mut(uri);
+        if (obj == nullptr) {
+            obj = emplace(uri, args...);
```
Comment (xlqian) [inline obj_factory.h]: return emplace(uri, args...);

File: obj_factory.h [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+            return nullptr;
+        }
+        return obj_it->second;
+    }
+
+    const ObjType* operator[] (const Idx<ObjType>& idx) const {
+        if (idx.val >= vec.size()) {
+            return nullptr;
+        }
+        return vec[idx.val];
```
Comment (xlqian) [inline obj_factory.h]: vec[idx.val].get();

File: obj_factory.h [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+            return nullptr;
+        }
+        return obj_it->second;
+    }
+
+    ObjType* get_mut(const Idx<ObjType>& idx) {
+        if (idx.val >= vec.size()) {
+            return nullptr;
+        }
+        return vec[idx.val];
```
Comment (xlqian) [inline obj_factory.h]: vec[idx.val].get();

File: obj_factory.h (hunk 9 above)
Comment (pbougue) [inline obj_factory.h]: seems clearer to me with less return in the middle but...

Comment (antoine-de) [inline obj_factory.h]: I quite agree with @pbougue on this

File: obj_factory.h [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+/*
+ * Factory handling all collections for one object
+ * Object must inherit from navitia::type::Header
+ */
+template<typename ObjType>
+class ObjFactory {
+private:
+    typedef typename std::vector<std::unique_ptr<ObjType>> inner_vector;
+    typedef typename std::unordered_map<std::string, ObjType*> inner_map;
```
Comment (TeXitoi) [inline obj_factory.h]: Why typedef here and using after?

File: obj_factory.h [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    inner_map map;
+
+public:
+    using iterator = typename inner_vector::iterator;
+    using const_iterator = typename inner_vector::const_iterator;
+
+    template<typename ...Args>
+    ObjType* emplace(const std::string& uri, Args&& ...args) {
+        if (navitia::contains(map, uri)) {
+            throw(std::logic_error(std::string("In ") + typeid(*this).name() +
```
Comment (TeXitoi) [inline obj_factory.h]: I'm not shure this `typeid(*this).name()` will do a readable message...

File: obj_factory.h (hunk 9 above)
Comment (TeXitoi) [inline obj_factory.h]:
```c++
if (ObjType* obj = get_mut(uri)) {
    return obj;
} else {
    return emplace(uri, args...);
}
```

File: obj_factory.h [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+        }
+        return vec[idx.val];
+    }
+
+    ObjType* get_mut(const std::string& uri) {
+        auto obj_it = map.find(uri);
+        if (obj_it == std::end(map)) {
+            return nullptr;
+        }
+        return obj_it->second;
```
Comment (TeXitoi) [inline obj_factory.h]: `return find_or_default(map, uri);`

File: obj_factory.h [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+    ObjType* insert(const std::string& uri, ObjType&& obj) {
+        return emplace(uri, obj);
+    }
+
+    const ObjType* operator[] (const std::string& uri) const {
+        const auto obj_it = map.find(uri);
+        if (obj_it == std::end(map)) {
+            return nullptr;
+        }
+        return obj_it->second;
```
Comment (TeXitoi) [inline obj_factory.h]: find_or_default

File: obj_factory.h (hunk 13 above)
Comment (pbougue) [inline obj_factory.h]: It is not user centred, sure, but it still offers some hints.
If you have better I'm curious, but alternatives I could think of are expensive.

**Comment by pbougue** [comment]
tests added

**Comment by pbougue** [comment]
Tests on non-copyable added

File: tests/obj_factory_test.cpp [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    i_int.val = 2;
+    auto iInt_bis = i_int;
+    navitia::ObjFactory<HeaderInt> obj_factory;
+
+    obj_factory.emplace("Hubert"); //name dropping
+    BOOST_CHECK_EQUAL(obj_factory["Hubert"]->val, 0);
+    obj_factory.get_or_create("Hubert", i_int); //should not change value
+    BOOST_CHECK_EQUAL(obj_factory[navitia::Idx<HeaderInt>(0)]->val, 0);
+    obj_factory.get_or_create("Hubert")->val = 1; //should change value
+    BOOST_CHECK_EQUAL(obj_factory.get_mut(navitia::Idx<HeaderInt>(0))->val, 1);
```
Comment (antoine-de) [inline tests/obj_factory_test.cpp]: why get_mut here ?

File: tests/obj_factory_test.cpp [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+BOOST_AUTO_TEST_CASE(simple_obj_factory) {
+    auto i_int = HeaderInt();
+    i_int.val = 2;
+    auto iInt_bis = i_int;
+    navitia::ObjFactory<HeaderInt> obj_factory;
+
+    obj_factory.emplace("Hubert"); //name dropping
+    BOOST_CHECK_EQUAL(obj_factory["Hubert"]->val, 0);
+    obj_factory.get_or_create("Hubert", i_int); //should not change value
+    BOOST_CHECK_EQUAL(obj_factory[navitia::Idx<HeaderInt>(0)]->val, 0);
```
Comment (antoine-de) [inline tests/obj_factory_test.cpp]: not very important, but maybe you can add a check on the `size()` to show that nothing have been added.

File: tests/obj_factory_test.cpp [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+    auto i_int = HeaderUniquePtrInt(2);
+    obj_factory.insert("John", std::move(i_int));
+    BOOST_CHECK_EQUAL(*(obj_factory.get_mut("John")->val), 2);
+
+    obj_factory.emplace("JohnEmpty");
+    obj_factory.get_or_create("Bob", 0);
+    obj_factory.get_or_create("BobEmpty");
+    auto iInt_bis = HeaderUniquePtrInt(2);
+    obj_factory.insert("OpenData", std::move(iInt_bis));
+    *(obj_factory.get_mut("OpenData")->val) = 4;
```
Comment (antoine-de) [inline tests/obj_factory_test.cpp]: don't you want to check that value ?

File: tests/obj_factory_test.cpp (hunk 16 above)
Comment (pbougue) [inline tests/obj_factory_test.cpp]: To test that all getters are working. Did not try to challenge constness on all of them though... Seems enough to me.

**Comment by TeXitoi** [comment]
OK for me.

File: tests/obj_factory_test.cpp (hunk 17 above)
Comment (pbougue) [inline tests/obj_factory_test.cpp]: done

File: tests/obj_factory_test.cpp (hunk 18 above)
Comment (pbougue) [inline tests/obj_factory_test.cpp]: done

**Pull request merged**